### PR TITLE
resolve: Pre-compute non-reexport module children

### PR DESCRIPTION
**Pull request by petrochenkov**
Instead of repeating the same logic by walking HIR during metadata encoding.

The only difference is that we are no longer encoding `macro_rules` items, but we never currently need them as a part of this list. They can be encoded separately if this need ever arises.

`module_reexports` is also un-querified, because I don't see any reasons to make it a query, only overhead.

### CODE REVIEW

File: compiler/rustc_resolve/src/imports.rs [hunk 1]
```diff
@@ -1261,10 +1261,14 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
 
         if let Some(def_id) = module.opt_def_id() {
+            let mut non_reexports = Vec::new();
             let mut reexports = Vec::new();
 
             module.for_each_child(self, |this, ident, _, binding| {
-                if let Some(res) = this.is_reexport(binding) {
+                let res = binding.res().expect_non_local();
+                if !binding.is_import() {
+                    non_reexports.push(binding.res().def_id().expect_local());
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:1270]:
```suggestion
                    non_reexports.push(res.def_id().expect_local());
```

File: compiler/rustc_middle/src/ty/context.rs [hunk 2]
```diff
@@ -2459,6 +2460,14 @@ impl<'tcx> TyCtxt<'tcx> {
...
         }
     }
+
+    pub fn module_children_non_reexports(self, def_id: LocalDefId) -> &'tcx [LocalDefId] {
+        self.resolutions(()).module_children_non_reexports.get(&def_id).map_or(&[], |v| &v[..])
+    }
+
+    pub fn module_children_reexports(self, def_id: LocalDefId) -> &'tcx [ModChild] {
+        self.resolutions(()).module_children_reexports.get(&def_id).map_or(&[], |v| &v[..])
+    }
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:2470]: Could you document what those methods return? What is the distinction between reexport and non-reexport?

File: compiler/rustc_resolve/src/imports.rs [hunk 3]
```diff
@@ -1261,10 +1261,14 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
         if let Some(def_id) = module.opt_def_id() {
+            let mut non_reexports = Vec::new();
             let mut reexports = Vec::new();
 
             module.for_each_child(self, |this, ident, _, binding| {
-                if let Some(res) = this.is_reexport(binding) {
+                let res = binding.res().expect_non_local();
+                if !binding.is_import() {
+                    non_reexports.push(res.def_id().expect_local());
+                } else if res != def::Res::Err && !binding.is_ambiguity() {
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/imports.rs:1271]: For `Res::Err` and ambiguities, should this closure just be a no-op?

Comment (petrochenkov) [reply compiler/rustc_resolve/src/imports.rs:1271]: Yes, those are effectively imports that import nothing, at least from other crates' points of view.
(Although for ambiguities it's an open question - https://github.com/rust-lang/rust/issues/36837.)

**Pull request merged**